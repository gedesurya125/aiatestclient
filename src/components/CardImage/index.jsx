import React, { useState, useEffect } from "react";
// import { StyledCardImage } from './styledCardImage'
import { styled } from "@mui/material/styles";
import Card from "@mui/material/Card";
import CardHeader from "@mui/material/CardHeader";
import CardMedia from "@mui/material/CardMedia";
import CardContent from "@mui/material/CardContent";
import CardActions from "@mui/material/CardActions";
import Collapse from "@mui/material/Collapse";
import Avatar from "@mui/material/Avatar";
// import IconButton from "@mui/material/IconButton";
import Typography from "@mui/material/Typography";
import { blue } from "@mui/material/colors";
// import FavoriteIcon from "@mui/icons-material/Favorite";
// import ShareIcon from "@mui/icons-material/Share";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
// import MoreVertIcon from "@mui/icons-material/MoreVert";
import { findAuthor } from "../../tools/findAuthor";
import { formatDate } from "../../tools/dateReformat";
import { Button } from "@mui/material";
import TagsContainer from "./components/TagsContainer";

// const ExpandMore = styled((props) => {
//   const { expand, ...other } = props;
//   return <IconButton {...other} />;
// })(({ theme, expand }) => ({
//   transform: !expand ? "rotate(0deg)" : "rotate(180deg)",
//   marginLeft: "auto",
//   transition: theme.transitions.create("transform", {
//     duration: theme.transitions.duration.shortest,
//   }),
// }));

const ExpandMoreIconAnimated = styled((props) => {
  const { expand, ...other } = props;
  return <ExpandMoreIcon {...other} />;
})(({ theme, expand }) => ({
  transform: !expand ? "rotate(0deg)" : "rotate(180deg)",
  marginLeft: "auto",
  transition: theme.transitions.create("transform", {
    duration: theme.transitions.duration.shortest,
  }),
}));

const DescriptionContainer = styled("div")(({ theme }) => ({
  "& p:first-of-type":{
    textAlign: 'center',
    fontWeight: 'bold'
  },
  "& p:first-of-type > a":{
    textDecoration: 'none'
  },
  "& p > a > img": {
    // textAlign: 'center',
    width: "100%",
    height: "auto",
  },
}));

const CardImage = ({ imageDetails }) => {
  // console.log('INI IMAGE DETAILSNYA', imageDetails)
  const [expanded, setExpanded] = useState(false);
  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  const author = findAuthor(imageDetails.author);
  // console.log('ini authornya', author);


  useEffect(() => {
    return () => {
      setExpanded(false);
    }
  },[imageDetails])
  return (
    <Card sx={{ 
      width: {
        xs: '90%',
        sm: '43%',
        md: '23%'
      }
     }}>
      <CardHeader
        avatar={
          <Avatar sx={{ bgcolor: blue[500] }} aria-label="recipe">
            {author[0].toUpperCase()}
          </Avatar>
        }
        
        title={<Typography fontWeight="bold">{author}</Typography>}
        subheader={formatDate(imageDetails.date_taken)}
      />
      <CardMedia
        component="img"
        height="194"
        src={imageDetails.media.m}
        alt="Paella dish"
      />
      <CardContent>
        <Typography align="center">{imageDetails.title}</Typography>
      </CardContent>
      <CardActions disableSpacing>
        <Button
          size="small"
          fullWidth
          color="primary"
          onClick={handleExpandClick}
          endIcon={

              <ExpandMoreIconAnimated
                expand={expanded}
                aria-expanded={expanded}
                aria-label="show more"
              />
          }
        >
          Show More
        </Button>
      </CardActions>
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          <DescriptionContainer
            dangerouslySetInnerHTML={{ __html: imageDetails.description }}
          ></DescriptionContainer>
          <Typography marginTop="1em" align="center" fontWeight="bold" gutterBottom>Tags:</Typography>
          <TagsContainer tagArray={imageDetails.tags.split(' ')}/>
        </CardContent>
      </Collapse>
    </Card>
  );
};

export default CardImage;
