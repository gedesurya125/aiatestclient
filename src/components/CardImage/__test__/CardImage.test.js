import React from "react";
import reactDOM from "react-dom";
import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import CardImage from "../index";

const mockImageDetails = {
  title: "Mmm",
  link: "https://www.flickr.com/photos/193445720@N03/51634513532/",
  media: {
    m: "https://live.staticflickr.com/65535/51634513532_a11d09dfa7_m.jpg",
    date_taken: "2021-10-28T15:42:46-08:00",
  },
  description: ` <p><a href="https://www.flickr.com/people/193445720@N03/">oitilkte61</a> posted a photo:</p> <p><a href="https://www.flickr.com/photos/193445720@N03/51634513532/" title="Mmm"><img src="https://live.staticflickr.com/65535/51634513532_a11d09dfa7_m.jpg" width="180" height="240" alt="Mmm" /></a></p> `,
  published: "2021-10-28T14:44:27Z",
  author: 'nobody@flickr.com ("oitilkte61")',
  author_id: "193445720@N03",
  tags: "",
};

afterEach(cleanup)

it("render component without crashing", () => {
  const div = document.createElement("div");
  reactDOM.render(<CardImage imageDetails={mockImageDetails}/>, div);
  reactDOM.unmountComponentAtNode(div);
});

it("render component properly with action button show more clicked", () => {
  const {getByTestId } = render(<CardImage imageDetails={mockImageDetails}/>);
  fireEvent.click(getByTestId('show-more-button'));
  expect(getByTestId("card-description")).not.toBeEmptyDOMElement();
});

it("Matches snapshot 1", () => {
  const tree = renderer.create(<CardImage imageDetails={mockImageDetails}/>).toJSON();
  expect(tree).toMatchSnapshot();
})
