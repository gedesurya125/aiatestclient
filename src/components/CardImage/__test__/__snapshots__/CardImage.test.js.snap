// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-cqbc8s-MuiPaper-root-MuiCard-root"
  data-testid="image-card"
>
  <div
    className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
    >
      <div
        aria-label="recipe"
        className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-4vb5ob-MuiAvatar-root"
      >
        O
      </div>
    </div>
    <div
      className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
      >
        oitilkte61
      </p>
      <span
        className="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
      >
        December NaN, NaN
      </span>
    </div>
  </div>
  <img
    alt="Paella dish"
    className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
    height="194"
    src="https://live.staticflickr.com/65535/51634513532_a11d09dfa7_m.jpg"
  />
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-1uwgr7b-MuiTypography-root"
    >
      Mmm
    </p>
  </div>
  <div
    className="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-fullWidth MuiButtonBase-root css-ua5ybo-MuiButtonBase-root-MuiButton-root"
      data-testid="show-more-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Show Detail
      <span
        className="MuiButton-endIcon MuiButton-iconSizeSmall css-155nyw6-MuiButton-endIcon"
      >
        <svg
          aria-expanded={false}
          aria-hidden={true}
          aria-label="show more"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1lcim9z-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
